autotelic/fastify-hmac
# Feature/add request decorator

resolves #12 
### Summary
- Refactored Plugin registration function to add a request decorator instead of adding a hook
- Updated base example to utilize the plugin as a `preValidation` hook.
- Added new example to demonstrate using the plugin request decorator in a route level `preValidation` hook.
- Updated Readme to show the different use cases.

### Test Plan
- run global hook example `npm run example:hook -- -l info -w`
  - Verify that requests with valid Signature headers are appropriately validated by the plugin.
  - Change request header and confirm that the plugin returns the verification error message.
  - Example HTTPie request:
 ```sh
http POST localhost:3000 Signature:'keyId="key-a", headers="(request-target) (created) (expires) host", signature="Nstb3q49bne90WmkbiZ9eyRtKCUmOWvc/kFyw/ftfLtqAXfRZGBk+hSJkXs0GRRm4Q/58c/pdIKj5DND11/fwQ==", created=1402170695, expires=1402170895'
```
- run route level decorator example `npm run example:decorator -- -l info -w`
  - Verify that any request sent to `/` is passed and the fastify server responds with `{ hello: 'no validation needed'}`.
  - Verify that requests with valid Signature headers sent to /foo are validated by the plugin. Fastify server will respond with `{ hello: 'validated HMAC' }`.
  - Change request header sent to `/foo` and confirm that the plugin returns the verification error message.
  - Example HTTPie request:
```sh
http POST localhost:3000/foo Signature:'keyId="key-a", headers="(request-target) (created) (expires) host", signature="fKj7/7DbxLi1RkMqBrPB6HJ6dXUThRM4nwSErMo3ni1SuXZjTEpNDiZozUMUYqpcGssakLQl2y8ryBz0U5oxlg==", created=1402170695, expires=1402170895'
```
- **NOTE:** Any implementation currently using the `fastify-hmac` plugin will need to be updated to use this new request decorator pattern.
